Ensure RemoteScrollbarsController is used by scrollable areas with async scrolling

nmoucht opened the pull request:
#### 5b9a2467dd21e23e6ac66ee621f68b35611b0f6d
<pre>
Ensure RemoteScrollbarsController is used by scrollable areas with async scrolling
<a href="https://bugs.webkit.org/show_bug.cgi?id=272935">https://bugs.webkit.org/show_bug.cgi?id=272935</a>
<a href="https://rdar.apple.com/126719107">rdar://126719107</a>

Reviewed by Simon Fraser.

After <a href="https://commits.webkit.org/277015@main">https://commits.webkit.org/277015@main</a> added use of usesAsyncScrolling to determine the scrollbars
controller type, and since usesAsyncScrolling is not correct prior to layer creation, ensure creation of
a RemoteScrollbarsController after layer creation if the scrollable area does use async scrolling.

* Source/WebCore/platform/ScrollableArea.h:
(WebCore::ScrollableArea::updateScrollbarsControllerForLayerCreation):
(WebCore::ScrollableArea::hasRemoteScrollbarsController):
* Source/WebCore/platform/ScrollbarsController.h:
(WebCore::ScrollbarsController::isRemoteScrollbarsController):
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::updateScrollingNodeForScrollingRole):
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::updateScrollbarsControllerForLayerCreation):
(WebCore::RenderLayerScrollableArea::createScrollbarsController):
* Source/WebCore/rendering/RenderLayerScrollableArea.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::createScrollbarsController const):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.h:

Canonical link: <a href="https://commits.webkit.org/278517@main">https://commits.webkit.org/278517@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bbaeffd591d3d2f1a2b9a3aa6ed35cd357652dec

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50711 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/30008 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/3028 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53970 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/1402 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/53014 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/36270 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/1052 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/41325 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52810 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/27657 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43684 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22447 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/25028 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/940 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/9152 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/47009 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/1002 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55560 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25813 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/894 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/48738 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/27070 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43804 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/47801 "Found 1 new API test failure: /WebKitGTK/TestWebKitWebContext:/webkit/WebKitWebContext/proxy (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11131 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/27938 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/26802 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on LayoutTests/fast/scrolling/mac/scrollbars/scrollbars-controller-type.html:
```diff
@@ -0,0 +1,67 @@
+
+            <div id="svgScroller" class="scroller">
+                <div class="contents">
+                    Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proid ...
```
You don't really need this text here.

nmoucht replied:
Removed.

smfr commented on LayoutTests/fast/scrolling/mac/scrollbars/scrollbars-controller-type.html:
```diff
@@ -0,0 +1,67 @@
+
+        async function doTest()
+        {
+            description('Test hover over the overlay scrollbar of a overflow: scroll in an SVG foreign object');
```
It would be good if the test name reflected what it was testing (like this) rather than some internal detail about the controller type.

nmoucht replied:
Updated this description as it wasn't accurate. We need to test this internal detail as there isn't a good way of detecting a user visible issue from it.

smfr commented on Source/WebCore/platform/ScrollableArea.h:
```diff
@@ -424,6 +425,8 @@ class ScrollableArea : public CanMakeWeakPtr<ScrollableArea> {
     virtual void updateScrollAnchoringElement() { }
     virtual void updateScrollPositionForScrollAnchoringController() { }
     virtual void invalidateScrollAnchoringElement() { }
+    virtual void updateScrollbarsControllerForLayerCreation() { }
```
I'm not sure what the "for layer creation" implies here.

nmoucht replied:
Changed this to be `updateScrollbarsControllerAfterLayerCreation`.

smfr commented on Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
```diff
@@ -845,14 +845,22 @@ bool RenderLayerScrollableArea::canShowNonOverlayScrollbars() const
 {
-    if (usesAsyncScrolling()) {
+    if (!hasRemoteScrollbarsController()) {
         if (auto scrollbarController = m_layer.page().chrome().client().createScrollbarsController(m_layer.page(), *this)) {
```
Why not just call `createScrollbarsController()` here?

Testing `! hasRemoteScrollbarsController()` then calling `chrome().client().createScrollbarsController()` implies some knowledge that it's the client that creates the `RemoteScrollbarsController`, but this code shouldn't have to know about that.

Instead this should do something like ask the client "does this RenderLayerScrollableArea have the right type of scrollbars controller already?"

nmoucht replied:
Right yeah this should be the other way around.

smfr commented on Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
```diff
@@ -845,13 +845,19 @@ bool RenderLayerScrollableArea::canShowNonOverlayScrollbars() const
 
+void RenderLayerScrollableArea::updateScrollbarsControllerAfterLayerCreation()
+{
+    if (auto scrollbarController = m_layer.page().chrome().client().createScrollbarsController(m_layer.page(), *this)) {
```
You've still got some "implicit" handshakes in this code; `WebChromeClient::createScrollbarsController()` checks whether the ScrollableArea does async scrolling, and returns null. So someone implementing `createScrollbarsController()` in their own chrome client would have to know that there's a "return null means no change" contract here.

It's also weird for ScrollableArea to expose an implementation detail (`hasRemoteScrollbarsController()`).

I think a better approach would be to have ChromeClient implement:
`updateScrollbarsControllerIfNecessary()`, and have it get the `ScrollbarsController` from the `ScrollableArea`, and then ask about its type.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
```diff
@@ -845,13 +845,19 @@ bool RenderLayerScrollableArea::canShowNonOverlayScrollbars() const
     return canHaveScrollbars() && !(m_layer.renderBox() && m_layer.renderBox()->canUseOverlayScrollbars());
 }
 
+void RenderLayerScrollableArea::updateScrollbarsControllerAfterLayerCreation()
```
This could be called `updateScrollbarsControllerIfNecessary()` since it's not specifically after layers have been newly created.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
```diff
@@ -845,13 +845,19 @@ bool RenderLayerScrollableArea::canShowNonOverlayScrollbars() const
-            return;
-        }
+    if (auto scrollbarController = m_layer.page().chrome().client().createScrollbarsController(m_layer.page(), *this)) {
+        setScrollbarsController(WTFMove(scrollbarController));
```
Try to avoid duplicating this code in 2 places.

nmoucht replied:
Moved to `updateScrollbarsControllerIfNecessary`.

smfr commented on Source/WebCore/platform/ScrollbarsController.h:
```diff
@@ -109,6 +110,8 @@ class ScrollbarsController {
 
     WEBCORE_EXPORT virtual void updateScrollbarStyle() { }
 
+    WEBCORE_EXPORT virtual bool isRemoteScrollbarsController() { return false; }
```
You should add `SPECIALIZE_TYPE_TRAITS` support on ScrollbarsController, so that the caller can just use `is<RemoteScrollbarsController>()`

nmoucht replied:
Fixed.

smfr commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
```diff
@@ -298,6 +298,7 @@ class WebChromeClient final : public WebCore::ChromeClient {
 
 #if PLATFORM(MAC)
     std::unique_ptr<WebCore::ScrollbarsController> createScrollbarsController(WebCore::Page&, WebCore::ScrollableArea&) const final;
```
Would be nice to avoid both `create` and `update` functions. Maybe just an `ensure` function that does both?

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/platform/ScrollbarsController.h:
```diff
@@ -109,9 +110,16 @@ class ScrollbarsController {
 
 } // namespace WebCore
+
+#define SPECIALIZE_TYPE_TRAITS_IS_REMOTE_SCROLLBARS_CONTROLLER(ToValueTypeName, predicate) \
```
Here in the base class the macros should not refer to the derived class. So this should be
`SPECIALIZE_TYPE_TRAITS_SCROLLBARS_CONTROLLER`

smfr replied:
Also consider if you should add a `SPECIALIZE_TYPE_TRAITS_SCROLLBARS_CONTROLLER` to ScrollbarsControllerMac, ScrollbarsControllerMock and any other subclasses, which would allow you to write code like `is<ScrollbarsControllerMock()...`

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/platform/ScrollableArea.h:
```diff
@@ -215,6 +216,7 @@ class ScrollableArea : public CanMakeWeakPtr<ScrollableArea> {
     ScrollAnimator* existingScrollAnimator() const { return m_scrollAnimator.get(); }
 
     WEBCORE_EXPORT ScrollbarsController& scrollbarsController() const;
+    ScrollbarsController* rawScrollbarsController() const { return m_scrollbarsController.get(); }
```
"raw" is unusual; we don't do that anywhere else. Use `existingScrollbarsController()`

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
```diff
@@ -847,14 +847,9 @@ bool RenderLayerScrollableArea::canShowNonOverlayScrollbars() const
-    }
-
-    ScrollableArea::createScrollbarsController();
+    m_layer.page().chrome().client().ensureScrollbarsController(m_layer.page(), *this);
```
If the client has a function called `ensureScrollbarsController()` then it should always make one, which means the default client implementation has to fall back to calling `ScrollableArea::createScrollbarsController()`. Or you call it something like `adjustScrollbarsController()` and allow it to leave the scrollbars controller null or null it out if the type is wrong.

nmoucht replied:
Right fixed.

smfr commented on Source/WebCore/platform/ScrollbarsController.h:
```diff
@@ -92,7 +92,8 @@ class ScrollbarsController {
     WEBCORE_EXPORT virtual String horizontalScrollbarStateForTesting() const { return emptyString(); }
     WEBCORE_EXPORT virtual String verticalScrollbarStateForTesting() const { return emptyString(); }
-    
+    WEBCORE_EXPORT virtual String scrollbarsControllerTypeForTesting() const { return emptyString(); }
```
If you add type trait support to all the subclasses, you don't need this; you could have `Internals::scrollbarsControllerTypeForNode` do the `is<>` checks to return the right type string.

nmoucht replied:
Fixed.

smfr commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
```diff
@@ -1171,22 +1171,29 @@ RefPtr<WebCore::ScrollingCoordinator> WebChromeClient::createScrollingCoordinato
-    if (area.mockScrollbarsControllerEnabled())
-        return nullptr;
+    auto* currentScrollbarsController = area.rawScrollbarsController();
+    if (area.mockScrollbarsControllerEnabled() || (currentScrollbarsController && is<RemoteScrollbarsController>(currentScrollbarsController)))
```
It's only correct to keep the existing `RemoteScrollbarsController` if `area.usesAsyncScrolling()` is still true, right? If a ScrollableArea goes from async to non-async, we have to change the ScrollbarsController back.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -3386,6 +3387,29 @@ ExceptionOr<String> Internals::verticalScrollbarState(Node* node) const
+{
+    if (is<ScrollbarsControllerMac>(controller))
+        return "ScrollbarsControllerMac"_s;
+    return "RemoteScrollbarsController"_s;
```
Why #ifdef for PLATFORM(MAC)? This should just check for all the subclasses, and work for the mock controller.

nmoucht replied:
Added check for mock controller. Added the ifdef just for ScrollbarsControllerMac.

smfr commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
```diff
@@ -1173,22 +1173,29 @@ RefPtr<WebCore::ScrollingCoordinator> WebChromeClient::createScrollingCoordinato
     ASSERT(page->corePage() == &corePage);
-    
+    auto* currentScrollbarsController = area.existingScrollbarsController();
     if (area.mockScrollbarsControllerEnabled())
```
Should this assert that we've already created a mock scrollbars controller, or does that happen later?

nmoucht replied:
Added assert(). We should never get into this code if mock scrollbars are enabled but adding it as a sanity check is fine.

nmoucht committed:
Ensure RemoteScrollbarsController is used by scrollable areas with async scrolling
https://bugs.webkit.org/show_bug.cgi?id=272935
rdar://126719107

Reviewed by Simon Fraser.

After https://commits.webkit.org/277015@main added use of usesAsyncScrolling to determine the scrollbars
controller type, and since usesAsyncScrolling is not correct prior to layer creation, ensure creation of
a RemoteScrollbarsController after layer creation if the scrollable area does use async scrolling.

* Source/WebCore/platform/ScrollableArea.h:
(WebCore::ScrollableArea::updateScrollbarsControllerForLayerCreation):
(WebCore::ScrollableArea::hasRemoteScrollbarsController):
* Source/WebCore/platform/ScrollbarsController.h:
(WebCore::ScrollbarsController::isRemoteScrollbarsController):
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::updateScrollingNodeForScrollingRole):
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::updateScrollbarsControllerForLayerCreation):
(WebCore::RenderLayerScrollableArea::createScrollbarsController):
* Source/WebCore/rendering/RenderLayerScrollableArea.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::createScrollbarsController const):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.h:

Canonical link: https://commits.webkit.org/278517@main
